### ISSUE TITLE
EnumeratePaths via a AggregateFileSystem is slower than it should be

### ISSUE THREAD

**Issue by garrynewman** [xoofx/zio#57]
### Problem

`EnumeratePaths` via `AggregateFileSystem` is searching folder by folder when using `SearchOption.AllDirectories` 

### Description

I've been hitting a problem in my codebase for a while where doing a simple EnumeratePaths on a AggregateFileSystem with a couple of Physical subsystems was very slow, taking just over **10 seconds** to find 8000 files on my Samsung 980 NVME SSD. 

This is one of the folders I'm enumerating over.

![image](https://user-images.githubusercontent.com/928843/119615781-afdcff80-bdf7-11eb-90ea-bc52094eb6d4.png)

So obviously I thought this is probably pretty expected and was about to write a bunch of caching stuff, but when looking at the code I feel like way [AggregateFileSystem is doing stuff here is bad for everyone](https://github.com/xoofx/zio/blob/87c7172ff63f4fb4785af2f0437485be7043c80e/src/Zio/FileSystems/AggregateFileSystem.cs#L338). 

### Expected

I expected it to call EnumeratePaths on all of the child filesystems and return a list of distinct results. There's probably a good reason it works this way instead, but I can't think of it.

**Comment by xoofx** [comment]
An aggregate file system is merging multiple file systems into one. It means that 2 filesystems inside an aggregate can have a same folder with same (name) and different (name) files and would have to return the list that makes sense in the order the aggregate. 
Hence why the code is going through the list of the filesystems AND performing a `entries.Contains(item))` because it has to take the first filesystem seen in the aggregate.

> I expected it to call EnumeratePaths on all of the child filesystems and return a list of distinct results. There's probably a good reason it works this way instead, but I can't think of it.

Maybe the name is misleading, but that's not what the Aggregate filesystem is doing (think about it as a layered file systems maybe?). What you are looking for is a different filesystem that I don't know how it would work (e.g if you have distinct results, you are supposed to find the file individually then on the aggregate, how would that work?)

Not saying that the `AggregateFileSystem` couldn't be optimized but the behavior of the code you linked above is intended.

**Comment by garrynewman** [comment]
Maybe I'm using the word distinct wrong here, If two child filesystems had the same file in the same folder, I'd expect it to return only one result for it, as if the filesystem was flattened and files had been overwritten.

I think my issue is really with it looping through each folder - that's what I don't understand. Wouldn't you get the same result by getting a list of paths from this kind of thing:

```
foreach ( child in children_filesystems )
{
    var paths = child.EnumeratePaths( same arguments as caller );

   foreach ( path in paths )
   {
      if ( already returned path ) continue
      yield return path;
   }
}
```

**Comment by xoofx** [comment]
Fixed by issue #58
